crates/ruff_linter/src/settings/types.rs

```diff
@@ -296,13 +296,22 @@ impl CacheKey for FilePatternSet {
 pub struct PerFileIgnore {
     pub(crate) basename: String,
     pub(crate) absolute: PathBuf,
+    pub(crate) negated: bool,
     pub(crate) rules: RuleSet,
 }
 
 impl PerFileIgnore {
-    pub fn new(pattern: String, prefixes: &[RuleSelector], project_root: Option<&Path>) -> Self {
+    pub fn new(
+        mut pattern: String,
+        prefixes: &[RuleSelector],
+        project_root: Option<&Path>,
+    ) -> Self {
         // Rules in preview are included here even if preview mode is disabled; it's safe to ignore disabled rules
         let rules: RuleSet = prefixes.iter().flat_map(RuleSelector::all_rules).collect();
+        let negated = pattern.starts_with('!');
+        if negated {
+            pattern.drain(..1);
+        }
```

Comment:
This might look a little odd, but it's zero-allocation, unlike the version using `.strip_prefix()`.

Perf probably doesn't matter much here in config-parsing; if you'd prefer the version that takes a non-mut String and uses `.strip_prefix()`, I can switch to that, though the code actually ends up longer than this.

Reply:
I'm cool with this.

Reply:
\cc @BurntSushi may have interesting things to say here :)

Reply:
I think this is fine personally. I agree that saving an alloc probably doesn't matter here (I assume these ultimately get converted to a glob matcher, and that is an enormous amount of work compared to saving an alloc) so I'd write whatever is clear. And this seems clear to me. :)